rust/src/operations/mod.rs

```diff
@@ -0,0 +1,287 @@
+        self.table.update().await?;
+
+        Ok(())
+    }
+
+    /// Create a new Delta table
+    pub async fn create(
+        &mut self,
+        metadata: DeltaTableMetaData,
+        mode: SaveMode,
+    ) -> DeltaCommandResult<()> {
+        let operation = DeltaOperation::Create {
+            mode,
+            metadata: metadata.clone(),
+            location: self.table.table_uri.clone(),
+            // TODO get the protocol from somewhere central
+            protocol: Protocol {
+                min_reader_version: 1,
+                min_writer_version: 2,
+            },
+        };
+        let plan = Arc::new(CreateCommand::try_new(
+            &self.table.table_uri,
+            operation.clone(),
+        )?);
+
+        self.execute(operation, plan).await
+    }
+
+    /// Write data to Delta table
+    pub async fn write(
+        &mut self,
+        data: Vec<RecordBatch>,
+        mode: SaveMode,
+        partition_columns: Option<Vec<String>>,
+    ) -> DeltaCommandResult<()> {
+        if data.is_empty() {
+            return Ok(());
+        }
+        let schema = data[0].schema();
+        let current_part = match self.table.update().await {
+            Ok(_) => {
+                let metadata = self.table.get_metadata()?;
+                if let Some(cols) = partition_columns {
+                    if cols != metadata.partition_columns {
+                        todo!("Schema updates not yet implemented")
+                    }
+                };
+                Some(metadata.partition_columns.clone())
+            }
+            _ => partition_columns,
+        };
+
+        let data = if let Some(ref cols) = current_part {
+            // TODO partitioning should probably happen in its own plan ...
+            let mut partitions: HashMap<String, Vec<RecordBatch>> = HashMap::new();
+            for batch in data {
+                let divided =
+                    divide_by_partition_values(schema.clone(), cols.clone(), &batch).unwrap();
+                for part in divided {
+                    let key = get_partition_key(cols, &part.partition_values)?;
+                    match partitions.get_mut(&key) {
+                        Some(batches) => {
+                            batches.push(part.record_batch);
+                        }
+                        None => {
+                            partitions.insert(key, vec![part.record_batch]);
+                        }
+                    }
+                }
+            }
+            partitions.into_values().collect::<Vec<_>>()
+        } else {
+            vec![data]
+        };
+
+        let operation = DeltaOperation::Write {
+            mode,
+            partition_by: current_part.clone(),
+            predicate: None,
+        };
+        let data_plan = Arc::new(MemoryExec::try_new(&data, schema, None)?);
+        let plan = Arc::new(WriteCommand::try_new(
+            &self.table.table_uri,
```

Comment:
This is something we can optimize as followups, but I think in the long run, it would be good to pass the table state here with the version so we won't need to perform a table load when executing the WriteCommand plan.